From 06371862d6b21101a4edc3e05234d06631ed5f40 Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Mon, 10 May 2021 17:15:05 +0200
Subject: [PATCH 117/954] feat(st-fmc): update the driver to support STM32MP25

This patch adds the support of STM32MP25 FMC2 IP:
 - the bit used to enable the IP has moved from BCR1 to CFGR
 - 4 chip select are available
 - New compatible strings have been added

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 9a976856ce7a48e3928eee00ab772ff40ef1756b feat(st-fmc): update the driver to support STM32MP25]
Change-Id: I3f2d40daee01715c7d1231df0741765abadbf92c
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
---
 drivers/st/fmc/stm32_fmc2_nand.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/drivers/st/fmc/stm32_fmc2_nand.c b/drivers/st/fmc/stm32_fmc2_nand.c
index 4f4c8280c2..1f06eebefc 100644
--- a/drivers/st/fmc/stm32_fmc2_nand.c
+++ b/drivers/st/fmc/stm32_fmc2_nand.c
@@ -26,13 +26,21 @@
 #define TIMEOUT_US_1_MS			1000U
 
 /* FMC2 Compatibility */
+#if STM32MP13 || STM32MP15
 #define DT_FMC2_EBI_COMPAT		"st,stm32mp1-fmc2-ebi"
 #define DT_FMC2_NFC_COMPAT		"st,stm32mp1-fmc2-nfc"
 #define MAX_CS				2U
+#endif
+#if STM32MP25
+#define DT_FMC2_EBI_COMPAT		"st,stm32mp25-fmc2-ebi"
+#define DT_FMC2_NFC_COMPAT		"st,stm32mp25-fmc2-nfc"
+#define MAX_CS				4U
+#endif
 #define MAX_BANK			5U
 
 /* FMC2 Controller Registers */
 #define FMC2_BCR1			0x00U
+#define FMC2_CFGR			0x20U
 #define FMC2_PCR			0x80U
 #define FMC2_SR				0x84U
 #define FMC2_PMEM			0x88U
@@ -48,6 +56,8 @@
 
 /* FMC2_BCR1 register */
 #define FMC2_BCR1_FMC2EN		BIT(31)
+/* FMC2_CFGR register */
+#define FMC2_CFGR_FMC2EN		BIT(31)
 /* FMC2_PCR register */
 #define FMC2_PCR_PWAITEN		BIT(1)
 #define FMC2_PCR_PBKEN			BIT(2)
@@ -660,7 +670,6 @@ static void stm32_fmc2_write_data(struct nand_device *nand,
 static void stm32_fmc2_ctrl_init(void)
 {
 	uint32_t pcr = mmio_read_32(fmc2_base() + FMC2_PCR);
-	uint32_t bcr1 = mmio_read_32(fmc2_base() + FMC2_BCR1);
 
 	/* Enable wait feature and NAND flash memory bank */
 	pcr |= FMC2_PCR_PWAITEN;
@@ -688,9 +697,13 @@ static void stm32_fmc2_ctrl_init(void)
 	pcr |= FMC2_PCR_TAR(FMC2_PCR_TAR_DEFAULT);
 
 	/* Enable FMC2 controller */
-	bcr1 |= FMC2_BCR1_FMC2EN;
+#if STM32MP13 || STM32MP15
+	mmio_setbits_32(fmc2_base() + FMC2_BCR1, FMC2_BCR1_FMC2EN);
+#endif
+#if STM32MP25
+	mmio_setbits_32(fmc2_base() + FMC2_CFGR, FMC2_CFGR_FMC2EN);
+#endif
 
-	mmio_write_32(fmc2_base() + FMC2_BCR1, bcr1);
 	mmio_write_32(fmc2_base() + FMC2_PCR, pcr);
 	mmio_write_32(fmc2_base() + FMC2_PMEM, FMC2_PMEM_DEFAULT);
 	mmio_write_32(fmc2_base() + FMC2_PATT, FMC2_PATT_DEFAULT);
-- 
2.53.0

